### PR TITLE
GAPI Fluid: SIMD optimization for sep filters of 5x5 kernel size (gaussBlur)

### PR DESCRIPTION
**Pull request by akhakim**
### This pullrequest changes

force_builders=Linux AVX2,Custom
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

### CODE REVIEW

**Comment by akhakim** [comment]
@dmatveev

**Comment by elatkin** [comment]
@akhakim  
Could you please clarify: how to try accuracy and performance tests for this new kernel?  
I have tried your branch on my PC, but it is not obvious from test names if 5x5 was tested.

File: modules/gapi/perf/cpu/gapi_imgproc_perf_tests_fluid.cpp [hunk 1]
```diff
@@ -58,7 +58,7 @@ INSTANTIATE_TEST_CASE_P(BlurPerfTestFluid, BlurPerfTest,
 INSTANTIATE_TEST_CASE_P(GaussianBlurPerfTestFluid, GaussianBlurPerfTest,
     Combine(Values(ToleranceFilter(1e-3f, 0.01).to_compare_f()),
             Values(CV_8UC1, CV_8UC3, CV_16UC1, CV_16SC1, CV_32FC1),
-            Values(3),                                     // TODO: add size=5, when kernel is ready
+            Values(3, 5),
```
Comment (akhakim) [inline modules/gapi/perf/cpu/gapi_imgproc_perf_tests_fluid.cpp]: @elatkin 
As you can see here, I add numeral 5 to instantiation's parameters of performance tests for GaussianBlur. So now all performance tests run for 3x3 and 5x5 kernel sizes according to our test API.

File: modules/gapi/test/cpu/gapi_imgproc_tests_fluid.cpp [hunk 2]
```diff
@@ -106,7 +106,7 @@ INSTANTIATE_TEST_CASE_P(gaussBlurTestFluid, GaussianBlurTest,
                                 Values(-1),
                                 Values(IMGPROC_FLUID),
                                 Values(ToleranceFilter(1e-3f, 0.01).to_compare_obj()),
-                                Values(3))); // add kernel size=5 when implementation is ready
+                                Values(3, 5))); // add kernel size=5 when implementation is ready
```
Comment (akhakim) [inline modules/gapi/test/cpu/gapi_imgproc_tests_fluid.cpp]: As you can see here, I also added numeral 5 to instantiation's parameters of accuracy tests for fluid's GaussianBlur. So now all accuracy tests run not only for 3x3 but also for 5x5 kernel size according to our test API.

File: modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp [hunk 3]
```diff
@@ -1464,18 +1481,576 @@ void run_sepfilter3x3_impl(DST out[], const SRC *in[], int width, int chan,  \
...
+        run_sepfilter5x5_any2char<noscale>(reinterpret_cast<uchar*>(out), in,
+            width, chan, kx, ky, border, scale, delta,
+            buf, y, y0);
+        return;
+    }
+#endif  // CV_SIMD
+
+    // reference code is quite fast for any-to-float case,
+    // but not for any-to-integral due to very slow rounding
+    run_sepfilter5x5_reference<noscale>(out, in, width, chan, kx, ky, border,
```
Comment (rgarnov) [inline modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp]: All this dispatcher logic can be unified for any kernel size if you add template parameter for kernel size. Moreover, I believe that almost all the code for 3x3 and 5x5 can be unified with such approach

Comment (akhakim) [reply modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp]: I will create a separate task for this according to the Dmitry's request.

Comment (rgarnov) [reply modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp]: Ok

**Comment by smirnov-alexey** [comment]
Looks OK

File: modules/gapi/src/backends/fluid/gfluidimgproc.cpp [hunk 4]
```diff
@@ -800,10 +808,28 @@ GAPI_FLUID_KERNEL(GFluidGaussBlur, cv::gapi::imgproc::GGaussBlur, true)
...
+        else
+            buffLen = ksize.height;
+
         buf[0] = ky + kysize;
-        buf[1] = buf[0] + length;
-        buf[2] = buf[1] + length;
+        for (int i = 1; i < buffLen; ++i)
+        {
+            buf[i] = buf[i-1] + length;
+        }
```
Comment (dmatveev) [inline modules/gapi/src/backends/fluid/gfluidimgproc.cpp]: I believe this changes are coming from fix #15874 right? Let's wait for that PR to be merged so than this one would only contain optimization-relevant changes

File: modules/gapi/src/backends/fluid/gfluidimgproc.cpp [hunk 5]
```diff
@@ -800,10 +808,28 @@ GAPI_FLUID_KERNEL(GFluidGaussBlur, cv::gapi::imgproc::GGaussBlur, true)
...
 
-        float *buf[3];
+        const int buffSize = 5;
+        float *buf[buffSize];
+
+        int buffLen = 0;
+        if (ksize.height >= buffSize)
+            buffLen = buffSize;
+        else
+            buffLen = ksize.height;
```
Comment (elatkin) [inline modules/gapi/src/backends/fluid/gfluidimgproc.cpp]: Probably:  
```c++
GAPI_Assert(ksize.height <= buffSize);
int buffLen = ksize.height;
```

File: modules/gapi/src/backends/fluid/gfluidimgproc.cpp [hunk 6]
```diff
@@ -800,10 +808,28 @@ GAPI_FLUID_KERNEL(GFluidGaussBlur, cv::gapi::imgproc::GGaussBlur, true)
...
+            buf[i] = buf[i-1] + length;
+        }
+
+        if (buffSize > buffLen)
+        {
+            for (int j = buffLen; j < buffSize; ++j)
+            {
+                buf[j] = nullptr;
+            }
+        }
```
Comment (elatkin) [inline modules/gapi/src/backends/fluid/gfluidimgproc.cpp]: No need in this cycle if you initialize `buf` with zeros, e.g.:  
```c++
float *buf[buffSize] = {0};
```

File: modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp [hunk 7]
```diff
@@ -1464,18 +1481,576 @@ void run_sepfilter3x3_impl(DST out[], const SRC *in[], int width, int chan,  \
...
+    {
+        r[n] = (y - y0 + n) % 5;  // previous, this, next rows
+    }
+
+    const int length = width * chan;
+    const int shift = chan;
+
+    // horizontal pass
+
+    int k0 = (y == y0) ? 0 : 3;
```
Comment (elatkin) [inline modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp]: Probably:
```c++
int k0 = (y == y0) ? 0 : 3;
```
The idea of `k0` is to decide if you need to process all **5** rows in the buffer or just the last one.  
Here you process only the last one if the previous rows were already done at previous passes.

File: modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp [hunk 8]
```diff
@@ -1464,18 +1481,576 @@ void run_sepfilter3x3_impl(DST out[], const SRC *in[], int width, int chan,  \
...
```
Comment (elatkin) [inline modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp]: I think **4**, not **3**:
```c++
int k0 = (y == y0) ? 0 : 4;
```

File: modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp [hunk 9]
```diff
@@ -1464,18 +1481,576 @@ void run_sepfilter3x3_impl(DST out[], const SRC *in[], int width, int chan,  \
...
+
+    const int length = width * chan;
+    const int shift = chan;
+
+    // horizontal pass
+    // full horizontal pass is needed only if the very 1st row in ROI is handled;
+    // for 2nd and further rows, it's enough to convolve only the
+    // "next" row - as we can reuse buffers from previous calls to
+    // this kernel (Fluid does rows consequently: y=y0, y0+1, ...)
+    int k0 = (y == y0) ? 0 : 3;
```
Comment (elatkin) [inline modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp]: Probably **4**, not **3**:
```C++
int k0 = (y == y0) ? 0 : 4;
```

File: modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp [hunk 10]
```diff
@@ -1464,18 +1481,576 @@ void run_sepfilter3x3_impl(DST out[], const SRC *in[], int width, int chan,  \
...
+    int shift = chan;
+
+    // horizontal pass
+
+    // full horizontal pass is needed only if very 1st row in ROI;
+    // for 2nd and further rows, it is enough to convolve only the
+    // "next" row - as we can reuse buffers from previous calls to
+    // this kernel (Fluid does rows consequently: y=y0, y0+1, ...)
+
+    int k0 = (y == y0) ? 0 : 3;
```
Comment (elatkin) [inline modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp]:
```c++
int k0 = (y == y0) ? 0 : kyLen-1;
```

**Pull request merged**